### ISSUE TITLE
Set comprehension + set union related `Expression of type "Set[Unknown | str]" cannot be assigned to return type "Set[str]"`

### ISSUE THREAD

**Issue by 098799** [microsoft/pyright#1549]
**Describe the bug**
Incorrect behavior when using set comprehensions in conjuction with set union.

**To Reproduce**
```
❯ pyright --version
pyright 1.1.61
```
```
❯ python --version
Python 3.9.1
```

Code:
```python
from typing import List, Optional, Set


class Foo:
    def __init__(self, kw_argument: Optional[List[str]] = None):
        self.kw_argument = kw_argument or []

    def set_comp_union(self) -> Set[str]:
        return {string for string in self.kw_argument} | {'a'}

    def set_comp(self) -> Set[str]:
        return set(self.kw_argument)

    def normal_set(self) -> Set[str]:
        return set(self.kw_argument) | {'a'}
```
yields `Expression of type "Set[Unknown | str]" cannot be assigned to return type "Set[str]"` in the case of `set_comp_union` method, but not `normal_set` or even the `set_comp` even though they are patently the same. 

Full output of terminal invocation of pyright:
```
Assuming Python platform Linux
Searching for source files
Found 1 source file
.../foo.py
  9:16 - error: Expression of type "Set[str | Any]" cannot be assigned to return type "Set[str]"
  TypeVar "_T" is invariant
    Type "str | Any" cannot be assigned to type "str" (reportGeneralTypeIssues)
1 error, 0 warnings 
Completed in 0.708sec
```

**Expected behavior**
No errors in the above code.

**VS Code extension or command-line**
cli

**Comment by 098799** [comment]
Noticed I'm running an ancient version of pyright for whatever reason. Sorry for littering, issue resolved after updating.
